docs/core.md

```diff
@@ -0,0 +1,361 @@
+# Core library
+
+The *Core library* implements common assertions about the exit code and output
+captured with the most recent `run()` command.
+
+
+## `flunk()`
+
+Display an error message and fail. This function provides a convenient way to
+report failure in arbitrary situations. You can use it to implement your own
+helpers when the ones available do not meet your needs. Other functions use it
+internally as well.
+
+```bash
+@test 'flunk()' {
+  flunk 'this test always fails'
+}
+```
+
+The message can also be specified on the standard input.
+
+```bash
+@test 'flunk() with pipe' {
+  echo 'this test always fails' | flunk
+}
+```
+
+This function always fails and simply outputs the given message.
+
+```
+this test always fails
+```
+
+
+## `assert()`
+
+Fail if the given expression evaluates to false.
+
+```bash
+@test 'assert()' {
+  local value=1
+  assert [ "$value" -eq 0 ]
+}
+```
+
+On failure the failed expression is displayed.
+
+```
+-- assertion failed --
+condition : [ 1 -eq 0 ]
+--
+```
+
+
+## `assert_equal()`
+
+Fail if the two parameters, expected and actual value respectively, do not
+equal.
+
+```bash
+@test 'assert_equal()' {
+  local expected='want'
+  local actual='have'
+  assert_equal "$expected" "$actual"
+}
+```
+
+On failure the expected and actual values are displayed.
+
+```
+-- values do not equal --
+expected : want
+actual   : have
+--
+```
+
+If either value is longer than one line both are displayed in *multi-line*
+format.
+
+
+## `assert_output()`
+
+Fail if `$output` does not equal the expected output.
+
+```bash
+@test 'assert_output()' {
+  run echo 'have'
+  local expected='want'
+  assert_output "$expected"
+}
+```
+
+The expected output can also be specified on the standard input.
+
+```bash
+@test 'assert_output() with pipe' {
+  run echo 'have'
+  local expected='want'
+  echo "$expected" | assert_output
+}
+```
+
+On failure the expected and actual outputs are displayed.
+
+```
+-- output differs --
+expected : want
+actual   : have
+--
+```
+
+If either value is longer than one line both are displayed in *multi-line*
+format.
+
+
+## `assert_success()`
+
+This function tests `$status` and additionally `$output` depending on the
+number of parameters.
+
+
+### `$status` test
+
+When no parameters are specified, fail if `$status` is not zero.
+
+```bash
+@test 'assert_success() status only' {
+  run bash -c "echo 'have'; exit 1"
+  assert_success
+}
+```
+
+On failure `$status` and `$output` are displayed.
+
+```
+-- command failed --
+status : 1
+output : have
+--
+```
+
+If `$output` is longer than one line, it is displayed in *multi-line* format.
+
+
+### Optional `$output` test
```

Comment:
Output comparisons in `assert_success` and `assert_failure` may be too limiting because they perform exact matches.

I'm curious if you've thought about removing this argument and adding a separate `assert_output` helper. In the short term, `assert_line` could fold into `assert_output` as an option (`assert_output -l 2 ...`), and in the long term `assert_output` could gain partial match and regex match options.